### PR TITLE
 Disable a debug log category if prefixed with '-' 

### PR DESCRIPTION
**Pull request by sickpig**
This would be particular useful when you want to have all categories enabled but one, e.g. now you can use debug=all,-selectcoins, especially when `bitcoind` is executed automatically, e.g. via the python functional tests suite.

### CODE REVIEW

**Review by awemany — APPROVED** [review]
Nit: rather commented-out code than dead code - i read `dead code` as unreachable code ...

**Review by awemany — APPROVED** [review]
Looks good, didn't run though.

**Comment by sickpig** [comment]
@awemany wrt dead code definition, you have a point. going to change the commit message.

File: src/util.h [hunk 1]
```diff
@@ -217,6 +217,12 @@ std::string LogGetLabel(uint64_t category);
  */
 std::string LogGetAllString();
 
+/**
+ * Get a string containin all enabled categories
```
Comment (awemany) [inline src/util.h]: Typo:  containin*g*

File: src/util.cpp [hunk 2]
```diff
@@ -173,6 +173,21 @@ std::string LogGetAllString()
     return ret;
 }
 
+std::string LogGetAllEnableCategoriesString()
```
Comment (awemany) [inline src/util.cpp]: This is very, very similar to `LogGetAllString()`. Can't these two be combined or just one of them used?

Also, the check `x.first == NONE` seems superfluous and confusing.

**Review by awemany — CHANGES_REQUESTED** [review]

File: src/util.h (hunk 1 above)
Comment (sickpig) [reply src/util.h]: fixed by removing the function :smile:

File: src/util.cpp (hunk 2 above)
Comment (sickpig) [reply src/util.cpp]: Add to `LogGetAllString` a functionality to show a string containing the string representation of active debug category separated by white spaces.

File: src/util.cpp [hunk 3]
```diff
@@ -185,15 +185,32 @@ void LogInit()
...
+
     for (string const &cat : categories)
     {
         category = boost::algorithm::to_lower_copy(cat);
+
+        // remove the category from the list of enables one
+        // if label is suffixed with a dash
+        bool toggle_flag = true;
+
+        if (category.length() > 0 && category.at(0) == '-')
```
Comment (awemany) [inline src/util.cpp]: The extra check for `length` is new, correct? When I read it initially, I was wondering whether one should do it. I guess it is better to have.

**Review by awemany — APPROVED** [review]
Looks good though didn't test.

**Review by awemany — APPROVED** [review]

**Comment by gandrewstone** [comment]
We need some qa or unit tests that exercise this new feature.  Please try normal stuff and also weird stuff like "-all", "-1", and just "-".
This feature makes the order of repeated key/values significant.  What if I did:
debug=all
debug=-net

will the order be preserved?

**Comment by sickpig** [comment]
@gandrewstone will do.

**Comment by awemany** [comment]
Good point on the order of repeated values. Maybe making repeated values illegal helps to avoid users shooting themselves in the foot?

I think this could actually happen easily and lead to hair-pulling moments like 'Why the fuck is my debug category disabled?!11?' (for example for a long config file that doesn't fit onto a single screen)

**Comment by sickpig** [comment]
@gandrewstone 

Test added, still not complete but I will add more tests today.

As you said already "order of repeated key/values significant", so to remove any ambiguity and to get a deterministic behavior the vector of options is sorted in descending (alphabetical) order and all the duplicates are removed.

**Comment by sickpig** [comment]
@gandrewstone PR is ready

File: src/unlimited.cpp [hunk 4]
```diff
@@ -2024,6 +2036,7 @@ static const CRPCCommand commands[] =
 #endif
     { "util",               "getaddressforms",        &getaddressforms,        true  },
     { "util",               "log",                    &setlog,                 true  },
+    { "util",               "getdebugcategories",     &getdebugcategories,     true  },
```
Comment (gandrewstone) [inline src/unlimited.cpp]: I think that you can do this (dump which categories are on or off) with the "log" rpc command.  

But if it turns out to be not redundant, I think we should call it "getlogcategories".  "debug" is a bad name for what logs to turn on and so we should only use it to maintain backwards compatibility.

Comment (sickpig) [reply src/unlimited.cpp]: good idea, will use log rather than a new rpc.

Comment (sickpig) [reply src/unlimited.cpp]: done

**Pull request merged**